include/oneapi/dpl/pstl/zip_view_impl.h

```diff
@@ -0,0 +1,388 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_ZIP_VIEW_IMPL_H
+#define _ONEDPL_ZIP_VIEW_IMPL_H
+
+#if _ONEDPL_CPP20_RANGES_PRESENT
+
+#include <ranges>
+#include <vector>
+#include <type_traits>
+
+#include "tuple_impl.h"
+
+namespace oneapi
+{
+namespace dpl
+{
+
+namespace ranges
+{
+
+template <bool C, typename... Views>
+concept all_forward = ( std::ranges::forward_range<std::conditional_t<C, const Views, Views>> && ... );
+
+template <bool C, typename... Views>
+concept all_bidirectional = ( std::ranges::bidirectional_range<std::conditional_t<C, const Views, Views>> && ... );
+
+template <bool C, typename... Views>
+concept all_random_access = ( std::ranges::random_access_range<std::conditional_t<C, const Views, Views>> && ... );
+
+template <typename... Rs>
+concept zip_is_common =
+    (sizeof...(Rs) == 1 && ( std::ranges::common_range<Rs> && ... ))
+    ||
+    (!(std::ranges::bidirectional_range<Rs> && ...) && (std::ranges::common_range<Rs> && ...))
+    ||
+    ((std::ranges::random_access_range<Rs> && ...) && (std::ranges::sized_range<Rs> && ...));
+
+template <bool Const, typename... Views>
+struct declare_iterator_category {};
+
+template <bool Const, typename... Views>
+    requires all_forward<Const, Views...>
+struct declare_iterator_category<Const, Views...> {
+    using iterator_category = std::input_iterator_tag;
+};
+
+template <std::ranges::input_range... Views>
+    requires ((std::ranges::view<Views> && ... ) && (sizeof...(Views) > 0))
+class zip_view : public std::ranges::view_interface<zip_view<Views...>> {
+public:
+    template <typename... Types>
+    using tuple_type = oneapi::dpl::__internal::tuple<Types...>;
+
+private:
+    template <typename _ReturnAdapter, typename _F, typename _Tuple, std::size_t... _Ip>
+    static decltype(auto)
+    apply_to_tuple_impl(_ReturnAdapter __tr, _F __f, _Tuple& __t, std::index_sequence<_Ip...>)
+    {
+        return __tr(__f(oneapi::dpl::__internal::get_impl<_Ip>()(__t))...);
```

Comment:
Does it makes sense to use `std::get` here instead? Would not it be more readable?